File: client/src/main/java/in/clayfish/printful/utils/CustomTypeAdapterFactory.java
package in.clayfish.printful.utils;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.TypeAdapter;
import com.google.gson.TypeAdapterFactory;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

/**
 * @author shuklaalok7
 * @since 29/12/2016
 */
class CustomTypeAdapterFactory<C> implements TypeAdapterFactory {

    private final Class<C> customizedClass;

    public CustomTypeAdapterFactory(Class<C> customizedClass) {
        this.customizedClass = customizedClass;
    }

    @SuppressWarnings("unchecked") // we use a runtime check to guarantee that 'C' and 'T' are equal
    public final <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        return type.getRawType() == customizedClass
                ? (TypeAdapter<T>) customizeClassAdapter(gson, (TypeToken<C>) type)
                : null;
    }

    /**
     * @param gson
     * @param type
     * @return
     */
    private TypeAdapter<C> customizeClassAdapter(Gson gson, TypeToken<C> type) {
        final TypeAdapter<C> delegate = gson.getDelegateAdapter(this, type);
        final TypeAdapter<JsonElement> elementAdapter = gson.getAdapter(JsonElement.class);
        return new TypeAdapter<C>() {
            @Override
            public void write(JsonWriter out, C value) throws IOException {
                JsonElement tree = delegate.toJsonTree(value);
                beforeToJson(value, tree);
                elementAdapter.write(out, tree);
            }

            @Override
            public C read(JsonReader in) throws IOException {
                JsonElement tree = elementAdapter.read(in);
                beforeToObject(tree);
                return delegate.fromJsonTree(tree);
            }
        };
    }

    /**
     * Override this to muck with {@code toSerialize} before it is written to the outgoing JSON stream.
     *
     * @param source
     * @param json
     */
    protected void beforeToJson(C source, JsonElement json) {
        JsonObject jsonObject = json.getAsJsonObject();
        for (Field field : source.getClass().getDeclaredFields()) {
            field.setAccessible(true);
            String oldFieldName = field.getName();

            if (LibUtils.checkUpperCase(oldFieldName) > -1) {
                String newFieldName = LibUtils.convertToPhpStyle(oldFieldName);
                try {
                    Class<?> type = field.getType();

                    if (String.class.isAssignableFrom(type)) {
                        jsonObject.addProperty(newFieldName, (String) field.get(source));
                    } else if (Number.class.isAssignableFrom(type)) {
                        jsonObject.addProperty(newFieldName, (Number) field.get(source));
                    } else if (Character.class.isAssignableFrom(type)) {
                        jsonObject.addProperty(newFieldName, (Character) field.get(source));
                    } else if (Boolean.class.isAssignableFrom(type)) {
                        jsonObject.addProperty(newFieldName, (Boolean) field.get(source));
                    } else {
                        throw new IllegalArgumentException("Cannot serialize it, please use a custom serializer.");
                    }
                    jsonObject.remove(oldFieldName);
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    /**
     * Override this to muck with {@code deserialized} before it parsed into the application type.
     *
     * @param json
     */
    protected void beforeToObject(JsonElement json) {
        JsonObject jsonObject = json.getAsJsonObject();
        Map<String, String> newFieldOldFieldMap = new HashMap<>();
        for (Map.Entry<String, JsonElement> entry : jsonObject.entrySet()) {
            String foundFieldName = entry.getKey();
            if (foundFieldName.contains("_")) {
                String newFieldName = LibUtils.convertToCamelCase(foundFieldName);
                newFieldOldFieldMap.put(foundFieldName, newFieldName);
            }
        }

        for (Map.Entry<String, String> entry : newFieldOldFieldMap.entrySet()) {
            jsonObject.add(entry.getValue(), jsonObject.get(entry.getKey()));
            jsonObject.remove(entry.getKey());
        }
    }
}

File: client/src/main/java/in/clayfish/printful/utils/CustomTypeAdapterFactory.kt
package `in`.clayfish.printful.utils

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.TypeAdapter
import com.google.gson.TypeAdapterFactory
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import java.io.IOException
import java.util.*
import kotlin.reflect.KClass

/**
 * Taken from http://stackoverflow.com/questions/11271375/gson-custom-seralizer-for-one-variable-of-many-in-an-object-using-typeadapter
 *
 * @author shuklaalok7
 * @since 29/12/2016
 */
internal open class CustomTypeAdapterFactory<C : Any>(private val customizedClass: KClass<C>) : TypeAdapterFactory {

    // we use a runtime check to guarantee that 'C' and 'T' are equal
    override fun <T> create(gson: Gson, type: TypeToken<T>): TypeAdapter<T>? =
            if (type.rawType == customizedClass) customizeClassAdapter(gson, type as TypeToken<C>) as TypeAdapter<T> else null

    /**
     * @param gson
     * @param type
     * @return
     */
    private fun customizeClassAdapter(gson: Gson, type: TypeToken<C>): TypeAdapter<C> {
        val delegate = gson.getDelegateAdapter(this, type)
        val elementAdapter = gson.getAdapter(JsonElement::class.java)
        return object : TypeAdapter<C>() {
            @Throws(IOException::class)
            override fun write(out: JsonWriter, value: C) {
                val tree = delegate.toJsonTree(value)
                beforeToJson(value, tree)
                elementAdapter.write(out, tree)
            }

            @Throws(IOException::class)
            override fun read(`in`: JsonReader): C {
                val tree = elementAdapter.read(`in`)
                beforeToObject(tree)
                return delegate.fromJsonTree(tree)
            }
        }
    }

    /**
     * Override this to muck with `toSerialize` before it is written to the outgoing JSON stream.
     *
     * @param source
     * @param json
     */
    protected open fun beforeToJson(source: C?, json: JsonElement?) {
        if (json == null || source == null) {
            return
        }
        val jsonObject = json.asJsonObject
        for (field in source.javaClass.declaredFields) {
            field.isAccessible = true
            val oldFieldName = field.name
            if (LibUtils.checkUpperCase(oldFieldName) > -1) {
                val newFieldName = LibUtils.convertToPhpStyle(oldFieldName)
                try {
                    val type = field.type
                    when {
                        type.isPrimitive -> when {
                            Boolean::class.javaPrimitiveType!!.isAssignableFrom(type) -> jsonObject.addProperty(newFieldName, field[source] as Boolean)
                            Char::class.javaPrimitiveType!!.isAssignableFrom(type) -> jsonObject.addProperty(newFieldName, field[source] as Char)
                            else -> jsonObject.addProperty(newFieldName, field[source] as Number)
                        }
                        String::class.java.isAssignableFrom(type) -> jsonObject.addProperty(newFieldName, field[source] as String)
                        Number::class.java.isAssignableFrom(type) -> jsonObject.addProperty(newFieldName, field[source] as Number)
                        Char::class.java.isAssignableFrom(type) -> jsonObject.addProperty(newFieldName, field[source] as Char)
                        Boolean::class.java.isAssignableFrom(type) -> jsonObject.addProperty(newFieldName, field[source] as Boolean)
                        else -> jsonObject.add(newFieldName, jsonObject[oldFieldName])
                    }

                    jsonObject.remove(oldFieldName)
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }
            }
        }
    }

    /**
     * Override this to muck with `deserialized` before it parsed into the application type.
     *
     * @param json
     */
    protected open fun beforeToObject(json: JsonElement?) {
        if (json == null || json.isJsonNull) return

        val jsonObject = json.asJsonObject
        val newFieldOldFieldMap: MutableMap<String, String> = HashMap()

        for ((foundFieldName) in jsonObject.entrySet())
        // This can be handled by gson automatically
            if (foundFieldName.contains("_"))
                newFieldOldFieldMap[foundFieldName] = LibUtils.convertToCamelCase(foundFieldName)

        for ((key, value) in newFieldOldFieldMap) {
            jsonObject.add(value, jsonObject[key])
            jsonObject.remove(key)
        }
    }

}
